protocols/autonatv2/src/client/handler/dial_request.rs

```diff
@@ -0,0 +1,342 @@
+        oneshot::Sender<
+            Result<
+                Stream,
+                StreamUpgradeError<<ReadyUpgrade<StreamProtocol> as OutboundUpgradeSend>::Error>,
+            >,
+        >,
+    >,
+}
+
+impl Handler {
+    pub(crate) fn new() -> Self {
+        Self {
+            queued_events: VecDeque::new(),
+            outbound: FuturesSet::new(DEFAULT_TIMEOUT, DEFAULT_MAXIMUM_CAPACITY),
+            queued_streams: VecDeque::default(),
+        }
+    }
+
+    fn perform_request(&mut self, req: DialRequest) {
+        let (tx, rx) = oneshot::channel();
+        self.queued_streams.push_back(tx);
+        self.queued_events
+            .push_back(ConnectionHandlerEvent::OutboundSubstreamRequest {
+                protocol: SubstreamProtocol::new(REQUEST_UPGRADE, ()),
+            });
+        if self
+            .outbound
+            .try_push(start_substream_handle(req, rx))
+            .is_err()
+        {
+            tracing::debug!("Dial request dropped, too many requests in flight");
+        }
+    }
+}
+
+impl ConnectionHandler for Handler {
+    type FromBehaviour = FromBehaviour;
+
+    type ToBehaviour = ToBehaviour;
+
+    type InboundProtocol = DeniedUpgrade;
+
+    type OutboundProtocol = ReadyUpgrade<StreamProtocol>;
+
+    type InboundOpenInfo = ();
+
+    type OutboundOpenInfo = ();
+
+    fn listen_protocol(&self) -> SubstreamProtocol<Self::InboundProtocol, Self::InboundOpenInfo> {
+        SubstreamProtocol::new(DeniedUpgrade, ())
+    }
+
+    fn poll(
+        &mut self,
+        cx: &mut Context<'_>,
+    ) -> Poll<
+        ConnectionHandlerEvent<Self::OutboundProtocol, Self::OutboundOpenInfo, Self::ToBehaviour>,
+    > {
+        if let Some(event) = self.queued_events.pop_front() {
+            return Poll::Ready(event);
+        }
+        if let Poll::Ready(m) = self.outbound.poll_unpin(cx) {
+            return Poll::Ready(ConnectionHandlerEvent::NotifyBehaviour(
+                ToBehaviour::TestCompleted(m.map_err(Error::Timeout).and_then(identity)),
+            ));
+        }
+        Poll::Pending
+    }
+
+    fn on_behaviour_event(&mut self, event: Self::FromBehaviour) {
+        match event {
+            FromBehaviour::PerformRequest(req) => {
+                self.perform_request(req);
+            }
+        }
+    }
+
+    fn on_connection_event(
+        &mut self,
+        event: ConnectionEvent<
+            Self::InboundProtocol,
+            Self::OutboundProtocol,
+            Self::InboundOpenInfo,
+            Self::OutboundOpenInfo,
+        >,
+    ) {
+        match event {
+            ConnectionEvent::DialUpgradeError(DialUpgradeError { error, .. }) => {
+                tracing::debug!("Dial request failed: {}", error);
+                match self.queued_streams.pop_front() {
+                    Some(stream_tx) => {
+                        if let Err(_) = stream_tx.send(Err(error)) {
+                            tracing::warn!("Failed to send stream to dead handler");
```

Comment:
This is not worth a `warn!`. It can only happen if stream upgrade times out or we drop it in favor of a new request. In both cases, this `warn` would just attract attention that the operator can't do anything about.

When writing logs, consider that most people will actually enable `info` and above for all their dependencies. A `warn` is something super critical so it should IMO only be used when something is _about to_ go horribly wrong.

Reply:
I would not log this at all to be honest.

Reply:
Ok removing.